### PR TITLE
WFLY-9768 Upgrade to Weld 3.0.3.Final and support CDI API multiversion

### PR DESCRIPTION
**Pull request by manovotn**
WFLY-9768 - https://issues.jboss.org/browse/WFLY-9768

This should be the last piece of puzzle to enable EE 8 preview for Weld/CDI.
It switches default behaviour to CDI 1.2 and adds dependency on CDI-API bridge (https://github.com/wildfly/cdi-api-bridge; already released) which can bring in missing CDI 2.0 bits needed by Weld.

If user enabled EE 8 preview, CDI 2.0 jar is used and bridge module is empty. If running in EE 7 mode, CDI 1.2 jar is used and bridge contains a jar with missing pieces of CDI 2.0 api.

As for testing, I verified full Weld TS in EE 8 mode with patched WFLY as well as CDI TCK 1.2 (default mode) and CDI TCK 2.0 (ee 8 mode). I haven't executed full WFLY TS, that I am leaving to this PR.

### CODE REVIEW

**Comment by manovotn** [comment]
@mkouba @stuartwdouglas please review

**Review by stuartwdouglas — APPROVED** [review]

File: servlet-feature-pack/src/main/resources/modules/system/layers/base/javax/enterprise/api/main/module.xml [hunk 1]
```diff
@@ -34,6 +34,16 @@
     </dependencies>
 
     <resources>
-        <artifact name="${javax.enterprise:cdi-api}"/>
+        <artifact name="javax.enterprise:cdi-api:1.2">
+            <conditions>
+                <property-not-equal name="ee8.preview.mode" value="true" />
+            </conditions>
+        </artifact>
+        <artifact name="javax.enterprise:cdi-api:2.0">
```
Comment (stuartwdouglas) [inline servlet-feature-pack/src/main/resources/modules/system/layers/base/javax/enterprise/api/main/module.xml]: I think the most recent version should still be managed by the pom.xml, and only hard code the legacy version, which will hopefully be removed soonish anyway.

Comment (manovotn) [reply servlet-feature-pack/src/main/resources/modules/system/layers/base/javax/enterprise/api/main/module.xml]: Hmm, current pom is set to CDI 1.2.
We could swap it around to 2.0 and then do `<artifact name="${javax.enterprise:cdi-api}">`
Is that what you mean?

Comment (mkouba) [reply servlet-feature-pack/src/main/resources/modules/system/layers/base/javax/enterprise/api/main/module.xml]: I originally thought we should use versions from pom.xml for both APIs but Stuart's proposal to only hard code version for legacy API is reasonable.

> We could swap it around to 2.0 and then do `<artifact name="${javax.enterprise:cdi-api}">`

@manovotn +1

**Comment by mkouba** [comment]
@manovotn Could you also add `ALLOW_OPTIMIZED_CLEANUP=true` here:
https://github.com/wildfly/wildfly/blob/master/weld/subsystem/src/main/java/org/jboss/as/weld/deployment/processors/WeldDeploymentProcessor.java#L324

WildFly should be ready for "improved" [WELD-2456](https://issues.jboss.org/browse/WELD-2456) - https://issues.jboss.org/browse/WFLY-9732 is done.

**Comment by kabir** [comment]
@manovotn @mkouba please @mention me when it has been adjusted and I will remove the label

**Comment by manovotn** [comment]
@kabir updated PR, you can now remove the label. I also added smoke test for CDI 2.0 functionality on your EE8-temp suite. It simply verifies that you can use async events in CDI (2.0 only).

Note: Latest WFLY core update (most likely candidate) broke some of the TCK [(CDI included)](https://jenkins.hosts.mwqe.eng.bos.redhat.com/hudson/view/TCK/view/TCK7-WildFly/job/tck7-wildfly-cdi/) so there are some TCK failures for now and I am unable to re-verify this locally :-/

**Comment by kabir** [comment]
Retest this please

**Comment by jamezp** [comment]
@manovotn @mkouba @stuartwdouglas The https://github.com/wildfly/cdi-api-bridge needs a LICENSE.txt and in the pom the `<name>Apache License 2.0</name>` should be used. The current requirement is it needs to match a name from https://spdx.org/licenses/. I don't think it will require a new release, just an update soonish.

**Review by jamezp — APPROVED** [review]

**Comment by manovotn** [comment]
Added line with CDI to [Getting Started Guide](https://github.com/wildfly/wildfly/pull/10832/files#diff-ed643d92a716a00b8994f96d553a9216R134)

**Pull request merged**